### PR TITLE
4017 creating a searchparameter with an expression referencing metasecurity will return error hapi 1118

### PR DESCRIPTION
**Pull request by MykolaMedynskyiSCDR**
What was done:
- Updating the logic leading to the validation of the SP expression format to accept FHIRPath starting with resource types, elements and datatype as per the [FHIRPath spec;](https://www.hl7.org/fhir/fhirpath.html#rules)
- Updating SP validation logic to no longer perform validation when create SP of version older than DSTU3;
- Removing tests that are no longer relevant;
- Added change log.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 1]
```diff
@@ -148,9 +148,9 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
 				}
 			}
 
-			if (!theContext.getVersion().getVersion().isEqualOrNewerThan(FhirVersionEnum.R4)) {
+			if (theContext.getVersion().getVersion().isOlderThan(FhirVersionEnum.DSTU3)) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:151]: addd code for dstu3

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 2]
```diff
@@ -148,9 +148,9 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
 				}
 			}
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:150]: add empty dstu2 loggic block

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 3]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
+							parseExpression(isUnique, theResource, expression, theContext);
 						}
 					}
+
 				}
 			} else {
-				if (!isUnique && theResource.getType() != Enumerations.SearchParamType.COMPOSITE && theResource.getType() != Enumerations.SearchParamType.SPECIAL && !REGEX_SP_EXPRESSION_HAS_PATH.matcher(expression).matches()) {
-					throw new UnprocessableEntityException(Msg.code(1120) + "SearchParameter.expression value \"" + expression + "\" is invalid");
-				}
+				parseExpression(isUnique, theResource, expression, theContext);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:180]: we will not be touching the R4 processing for now.  restore previous implementation.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 4]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
 				}
 			}
 
-			if (!theContext.getVersion().getVersion().isEqualOrNewerThan(FhirVersionEnum.R4)) {
-
-				// DSTU3 and below
+			FhirVersionEnum currentVersion = theContext.getVersion().getVersion();
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:151]: rename variable to fhirVersion would be more relevant to what the variable represent.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 5]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
+			else if (currentVersion.isEquivalentTo(FhirVersionEnum.DSTU3)) {
 				String[] expressionSplit = theSearchParamExtractor.split(expression);
 				for (String nextPath : expressionSplit) {
 					nextPath = nextPath.trim();
 
 					int dotIdx = nextPath.indexOf('.');
-					if (dotIdx == -1) {
-						throw new UnprocessableEntityException(Msg.code(1117) + "Invalid SearchParameter.expression value \"" + nextPath + "\". Must start with a resource name.");
-					}
+					String resourceName = (dotIdx == -1) ? nextPath : nextPath.substring(0, dotIdx);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:161]: Let's wrap the resourceName extraction into a method that will return an Optional<String>:
private Optional<String> extractResourceNameFromExpression(String theExpression){...}

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 6]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
 
-					String resourceName = nextPath.substring(0, dotIdx);
-					try {
-						theContext.getResourceDefinition(resourceName);
-					} catch (DataFormatException e) {
-						throw new UnprocessableEntityException(Msg.code(1118) + "Invalid SearchParameter.expression value \"" + nextPath + "\": " + e.getMessage());
-					}
+					if (theDaoConfig.isValidateSearchParameterExpressionsOnSave()) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:163]: check if our new optional has indeed a resource name as part of the if statement

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 7]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
 
-					if (theContext.getVersion().getVersion().isEqualOrNewerThan(FhirVersionEnum.DSTU3)) {
-						if (theDaoConfig.isValidateSearchParameterExpressionsOnSave()) {
+						try {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:165]: Until we understand what's going on with parsing an expression, we are not going to validate an expression that does not start with a resource.  remove the encapsulating try/catch.

**Review by epeartree — COMMENTED** [review]
the next step is to understand what is happening with the parsing.
i presume, and verification is required, that the following line will assert that the fhirpath in the nextPath can be applied to the temporaryInstance.  
`theContext.newFluentPath().evaluate(temporaryInstance, nextPath, IBase.class);`

while line 
`theContext.newFluentPath().parse(theExpression);`
SHOULD be parsing the expression and scream if it contains unknown elements/datatype.  we have to figure out why we can pass it garbage and all is good.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 8]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
-					String resourceName = nextPath.substring(0, dotIdx);
-					try {
-						theContext.getResourceDefinition(resourceName);
-					} catch (DataFormatException e) {
-						throw new UnprocessableEntityException(Msg.code(1118) + "Invalid SearchParameter.expression value \"" + nextPath + "\": " + e.getMessage());
-					}
+					if (theDaoConfig.isValidateSearchParameterExpressionsOnSave()) {
 
-					if (theContext.getVersion().getVersion().isEqualOrNewerThan(FhirVersionEnum.DSTU3)) {
-						if (theDaoConfig.isValidateSearchParameterExpressionsOnSave()) {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:171]: we have to figure out to what setting/variable is this flag referring too and if we need to apply it to R4 as well(most likely).

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java [hunk 9]
```diff
@@ -148,52 +148,51 @@ public static void validateSearchParam(SearchParameter theResource, FhirContext
...
+						try {
 							IBaseResource temporaryInstance = theContext.getResourceDefinition(resourceName).newInstance();
 							try {
 								theContext.newFluentPath().evaluate(temporaryInstance, nextPath, IBase.class);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoSearchParameterR4.java:168]: method 'newFluentPaht' is deprecated.  replace as per documentation

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4017-creating-a-searchparameter-with-an-expression-referencing-metasecurity-will-return-error-hapi-1118.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 4037
+jira: SMILE-5046
+title: "Previously, creating a DSTU3 SearchParameter with an expression that does not start with a resource type would thrown an error. This has been corrected."
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4017-creating-a-searchparameter-with-an-expression-referencing-metasecurity-will-return-error-hapi-1118.yaml:5]:
```suggestion
title: "Previously, creating a DSTU3 SearchParameter with an expression that does not start with a resource type would throw an error. This has been corrected."
```

**Review by tadgh — CHANGES_REQUESTED** [review]
Overall looks good minus the logical change of no longer validating DSTU2 and older SPs. This should still occur.

**Review by tadgh — APPROVED** [review]

**Pull request merged**